### PR TITLE
feat: fall detection in UI

### LINKED ISSUE #464: Implement fall detection icon in UI
**Is your feature request related to a problem? Please describe.**
Currently fall detections are rendered with a generic question mark icon. See attached screenshot. We need to replace that with a better visual representation.

**Describe the solution you'd like**
Replace questionmark material design icon with a fall alert  icon.

IMG_5028 2

### PR DESCRIPTION
**Pull request by vickywane**
This resolves [this](https://github.com/ambianic/ambianic-ui/issues/464)

This PR implements an update of the icon within the fall detection circle to better illustrate a detected fallng user.

### CODE REVIEW

File: src/views/Timeline.vue [hunk 1]
```diff
@@ -464,7 +464,7 @@ export default {
       if (inf.length > 0) {
         topLabel = inf[0].label
       }
-      let icon = 'mdi-crosshairs-question'
+      let icon = 'mdi-karate'
```
Comment (ivelin) [inline src/views/Timeline.vue:467]: @vickywane here is a fall icon:

https://iconify.design/icon-sets/tabler/fall.html

**Review by ivelin — CHANGES_REQUESTED** [review]

File: public/index.html [hunk 2]
```diff
@@ -26,6 +27,7 @@
       name="description"
       content="Ambianic UI progressive web application to manage Ambianic Edge devices"
     />
+    <script src="https://code.iconify.design/1/1.0.7/iconify.min.js"></script>
```
Comment (ivelin) [inline public/index.html:30]: @vickywane can we pull this icon lib as a local install dependency for improved response time? I've noticed remote dependencies tend to break lighthouse if the serving host has low response time. Pulling as a local dependency makes them part of the CDN propagation. Some libraries are already hosted on CDNs and its OK to link to the their CDN url. 

Looks like iconify has its own public CDN: https://docs.iconify.design/sources/api/hosting.html

However I don't find icnonify in popular js cdn networks: https://cdnjs.com/libraries?q=iconify

What are your thoughts?

File: src/views/Timeline.vue [hunk 3]
```diff
@@ -90,7 +91,7 @@
             >
               <v-row
                 class="fill-height ma-0"
-                align="start"
+                align="tart"
```
Comment (ivelin) [inline src/views/Timeline.vue:94]: @vickywane is this a typo `tart` ?

File: src/views/Timeline.vue [hunk 4]
```diff
@@ -13,6 +13,7 @@
     >
       <v-card
         class="mx-auto"
+        edsf
```
Comment (ivelin) [inline src/views/Timeline.vue:16]: @vickywane what is `edsf` style for?

File: src/views/Timeline.vue [hunk 5]
```diff
@@ -119,11 +120,22 @@
                     align="top"
                     class="font-weight-regular pa-4 ma-6 see-thru"
                   >
+                    <div v-if="sample.args.inference_meta.display === 'Fall Detection'">
```
Comment (ivelin) [inline src/views/Timeline.vue:123]: @vickywane how about adding a few test cases to cover correct display of icons based on timeline event content?

File: src/views/Timeline.vue [hunk 6]
```diff
@@ -464,7 +476,7 @@ export default {
       if (inf.length > 0) {
         topLabel = inf[0].label
       }
-      let icon = 'mdi-crosshairs-question'
+      let icon = 'mdi-karate'
```
Comment (ivelin) [inline src/views/Timeline.vue:479]: @vickywane does this still need to be here ?

Comment (ivelin) [reply src/views/Timeline.vue:479]: Should we pull icon choice into its own `ambianic-event-view` component outside of the timeline view page?

**Review by ivelin — COMMENTED** [review]
@vickywane added a few comments.

File: public/index.html (hunk 2 above)
Comment (vickywane) [reply public/index.html:30]: I think we can make do with now. The iconify CDN is quite widely used by a lot of projects and should have a good response time. It's also cached by the browser so you don't need to pull it each time. 

I tested the display without an internet connection

File: src/views/Timeline.vue (hunk 3 above)
Comment (vickywane) [reply src/views/Timeline.vue:94]: Typo

File: src/views/Timeline.vue (hunk 4 above)
Comment (vickywane) [reply src/views/Timeline.vue:16]: Typo

File: src/views/Timeline.vue (hunk 5 above)
Comment (vickywane) [reply src/views/Timeline.vue:123]: Adding a test case would mean that I have to connect to an edge device using a valid Peer ID, for the timeline to fetch fall events.

Comment (vickywane) [reply src/views/Timeline.vue:123]: Is there a way we can have a valid PeerID to use? 

That way I can boost the codecov report of this PR

File: src/views/Timeline.vue (hunk 6 above)
Comment (vickywane) [reply src/views/Timeline.vue:479]: I have removed the first icon value. The icon from Iconfiy would be used for the fall icon.

File: public/index.html (hunk 2 above)
Comment (ivelin) [reply public/index.html:30]: OK

File: src/views/Timeline.vue [hunk 7]
```diff
@@ -90,7 +90,6 @@
             >
               <v-row
                 class="fill-height ma-0"
-                align="start"
```
Comment (ivelin) [inline src/views/Timeline.vue:93]: @vickywane can you please explain the result of changing from `align` to `justify`? Can you attach examples of before and after that show clearly the difference. Thank you.

File: src/views/Timeline.vue (hunk 5 above)
Comment (ivelin) [reply src/views/Timeline.vue:121]: That is incorrect. We have discussed this topic before. When the focus is on testing a specific component logic, all moving parts not directly used in the tested logic can be isolated. The classic unit test vs integration test vs end to end test discussion. In this case we are looking at a vuetify component unit test.

In the pre-test section, pre-populate vuex with values for the various timeline event objects and then test that the icon component renders the correct icon values bases on the provided timeline event data object.

File: src/views/Timeline.vue [hunk 8]
```diff
@@ -464,7 +472,7 @@ export default {
       if (inf.length > 0) {
         topLabel = inf[0].label
       }
-      let icon = 'mdi-crosshairs-question'
+      let icon = ''
```
Comment (ivelin) [inline src/views/Timeline.vue:475]: @vickywane several comments here.

1. I see you placed the fall detection icon rendering logic in the vue markup instead of reusing and generalizing this `eventIcon` function that specifically picks the correct icon based on the event type. I see that you did it because the fall icon requires use of a third party library (iconify) instead of the standard MDI set. However the code now looks disorganized and messy.
2. Let's create an `<event-icon />` component that encapsulates the ugly details of rendering the correct icon markup. Move `eventIcon()` and the fall detection icon markup out of `timeline.vue` and replace with a simple and elegant one line `<event-icon :event-type="$sample.args.inference_meta.display" />` vue component reference.
3. You removed the use of 'mdi-crosshairs-question'. Is the intention to not show any event icon at all if the event type is uknown?

**Review by ivelin — CHANGES_REQUESTED** [review]

File: src/views/Timeline.vue (hunk 7 above)
Comment (vickywane) [reply src/views/Timeline.vue:93]: This was unintentional change

**Comment by ivelin** [comment]
@vickywane the codecov check fails because there is not a sufficient coverage of the differences between new code from this PR and old code prior to this PR. As the message states the target is at least 52.16%. Ideally 100%.:

`@codecov codecov/patch — 43.26% of diff hit (target 52.16%)`

If you follow the Details link it takes you to specific file names and line numbers that are not covered yet:
https://github.com/ambianic/ambianic-ui/pull/581/checks?check_run_id=2156197386
Screen Shot 2021-03-22 at 9 47 38 AM

Then if you click on any of the listed coverage cards, it takes you to the expanded source files with uncovered sections:

Screen Shot 2021-03-22 at 9 48 34 AM

Admittedly, the coverage before this PR was poor and you are doing extra work to catch us up with a higher quality standard. Thank you for that!

File: src/views/Timeline.vue [hunk 9]
```diff
@@ -356,88 +370,92 @@ export default {
...
-      try {
-        if (this.clearTimeline) {
-          this.timeline.length = 0
-          this.clearTimeline = false
-        }
-        const data = await this.getNextTimelinePage()
-        console.debug('Infinite handler received Bottom timeline page', { data }) // eslint-disable-line no-console
-        // Are there any more timeline events left?
-        if (data && data.timeline && data.timeline.length > 0) {
+      if (!window.Cypress) {
```
Comment (ivelin) [inline src/views/Timeline.vue:417]: @vickywane I see a lot of conditional logic that alters the code behavior if its executed in a test environment. That introduces a lot of risks to invalidate the test. Tests should be validating the actual code, not code specifically written for tests. Let's figure out a way to isolate components being tested and feeding them test data, rather than changing the logic of the code being tested.

**Review by ivelin — CHANGES_REQUESTED** [review]
@vickywane its good to see eventIcon is in its own component now and has proper component test coverage. However there are still a number of open questions. See comments.

File: src/views/Timeline.vue (hunk 9 above)
Comment (vickywane) [reply src/views/Timeline.vue:417]: Isolating the component leads to a much bigger issue due to the peer connection being made each time the peerConnectionStatus is changed to `CONNECTED`. 

The Vue recommended way would be mutate the stores from the test suite when tests are running, however the PWA makes background requests based on the `peerConnectionStatus` value and it’s seemingly impossible to isolate the component.

Take for instance the `timeline` component, it’s `created` method initiates a `edgeApi` Class which fetches the data from the connected edge device.

File: src/views/Timeline.vue [hunk 10]
```diff
@@ -4,7 +4,7 @@
     justify="space-around"
   >
     <v-col
-      v-if="!isEdgeConnected"
+      v-if="connectionStatus !== 'PEER_CONNECTED'"
```
[the remark this answers is not in the document]
Comment (vickywane) [reply src/views/Timeline.vue:7]: Instead of mutating the `edgeConnected` value directly in the store, i am copying the value from the store so i can simply access it as a data value in the component.

The goal for the tests is to detach the component or anything that relates to testing the component from the central VueX store

[the remark this answers is not in the document]
Comment (ivelin) [reply src/views/Timeline.vue:7]: Doesn't sound convincing. Let's think about encapsulating the tested code with a known pre-test state, running the test and verifying output against expected results. 

In this case, think about setting the value of isEdgeConnected prior to running a test and then verifying output against expected results.

File: src/views/Timeline.vue (hunk 9 above)
Comment (ivelin) [reply src/views/Timeline.vue:417]: Ok, let's work with your example

> Take for instance the timeline component, it’s created method initiates a edgeApi Class which fetches the data from the connected edge device.

If the goal is to test behaviour of `timeline` rendering for various inputs and `edgeAPI` is dynamically affecting the test outcomes, then we should think about ways to address that varying behaviour. One way to do this is to dynamically replace the definition of `created()` or `initEdgeAPI()` with known and realistic fixed values prior to executing the test and asserting against results. This way we reduce the test focus to a predictable set of pre-conditions and assert results based on these fixed pre-conditions without altering the production code that we are testing.

Respectively if the goal is to test the behavior of `created()` or `initEdgeAPI()` then you design tests that fix their inputs and assert on outputs.

**Review by ivelin — CHANGES_REQUESTED** [review]
@vickywane let's find a way to not modify production code. There is no much benefit in tests if they are not testing actual production code. Tests that are testing test-environment specific code are not ideal and they don't cover all the code that affects user experience.

File: src/views/Timeline.vue (hunk 6 above)
Comment (ivelin) [reply src/views/Timeline.vue]: OK. Good work moving that out into an `eventIcon` component.

File: src/views/Timeline.vue (hunk 5 above)
Comment (ivelin) [reply src/views/Timeline.vue]: > Is there a way we can have a valid PeerID to use?
> 
> That way I can boost the codecov report of this PR

Yes. Any UUID4 value would do. Just grab a fixed PeerID from somewhere and use it in tests.

**Comment by vickywane** [comment]
> @vickywane let's find a way to not modify production code. There is no much benefit in tests if they are not testing actual production code. Tests that are testing test-environment specific code are not ideal and they don't cover all the code that affects user experience.

This has now been properly corrected. The tests relating to the peer connection have been isolated and tested using Jest.

File: cypress/integration/ambianic-tests/timeline.spec.js [hunk 11]
```diff
@@ -1,29 +1,29 @@
 /// <reference types="cypress" />
 
 context('Timeline', () => {
-    before(() => {
```
Comment (ivelin) [inline cypress/integration/ambianic-tests/timeline.spec.js:4]: @vickywane please avoid committing changes that are not directly related to the PR. It adds time and confusion to the review process.

**Review by ivelin — CHANGES_REQUESTED** [review]
@vickywane please only commit the minimum changes required for this PR. Remove noisy commits.

**Comment by vickywane** [comment]
> @vickywane please avoid committing changes that are not directly related to the PR. It adds time and confusion to the review process.

I rebased against the master branch and picked the recent changes. The PR has been open for a significant while now so I had to make sure it was updated. 

The fall detection icon is also shown on the `Timeline` page.

File: cypress/integration/ambianic-tests/timeline.spec.js (hunk 11 above)
Comment (vickywane) [reply cypress/integration/ambianic-tests/timeline.spec.js:4]: The changes here are the same. I only linted the file before pushing it.

I have created [this feature request ](https://github.com/ambianic/ambianic-ui/issues/678)to provide a solution to the cause of the unneeded change when a linting is done.

File: src/views/Timeline.vue [hunk 12]
```diff
@@ -42,42 +43,30 @@
...
                 align="start"
                 justify="start"
               >
-                <template
-                  v-if="isImageLoaded[index]"
-                >
+                <template v-if="isImageLoaded[index]">
                   <detection-boxes
                     :detections="sample.args.inference_result"
+                    :tensor_image_size="
```
Comment (ivelin) [inline src/views/Timeline.vue:65]: @vickywane can you please elaborate on this attribute? Why was it added and how is it used?

Comment (ivelin) [inline src/views/Timeline.vue:65]: @vickywane can you please elaborate on this attribute? Why was it added and how is it used?

**Review by ivelin — CHANGES_REQUESTED** [review]

File: src/main.js [hunk 13]
```diff
@@ -18,13 +18,16 @@ const router = new VueRouter({
...
   router,
   store,
   vuetify,
   render: h => h(App)
-}).$mount('#app')
+})
+
+VueApp.$mount('#app')
 
 if (window.Cypress) {
```
Comment (ivelin) [inline src/main.js:30]: @vickywane is there a way to remove test framework dependencies from the tested code? Do we really need Cypress here? Instead, how about doing the environment setup as part of the test code in the test setup sections?

**Review by ivelin — CHANGES_REQUESTED** [review]
@vickywane this PR is not quite ready. See comments.

File: src/main.js (hunk 13 above)
Comment (vickywane) [reply src/main.js:30]: Since i refactored the test, there is no need for the Cypress instance here. 

I have restored the file back to it's original state.

File: src/components/utils.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
+export const setTimelineEventIcon = (inf) => {
+  let topLabel = 'none'
+
+  if (inf.length > 0) {
+    topLabel = inf[0].label
+  }
+  let icon = 'mdi-corshairs-question'
```
[the remark this answers is not in the document]
Comment (vickywane) [reply src/components/utils.js]: I have fixed this typo and also added a test case to the exisiting test cases.

File: src/views/Timeline.vue (hunk 12 above)
Comment (vickywane) [reply src/views/Timeline.vue]: I think it's from the `detection-boxes` as at the time when i created this branch. I have looked through the `DetectionBox` and `EventIcon` component and there is no use for it. 

I have removed it in my latest push.

File: src/main.js [hunk 15]
```diff
@@ -18,12 +18,14 @@ const router = new VueRouter({
...
```
Comment (ivelin) [inline src/main.js:30]: @vickywane the PR still tries to commit the cypress line. See above. Did you sync your local changes with the remote branch?

File: src/main.js (hunk 13 above)
Comment (ivelin) [reply src/main.js:30]: Seems like you forgot to synch your local changes with remote.

**Review by ivelin — CHANGES_REQUESTED** [review]
Cypress still in the deployed code. See comment.

**Comment by vickywane** [comment]
> Cypress still in the deployed code. See comment.

The cypress line is needed as we have a e2e test case for the `home.spec.js` file that dispatches an action to the store. 

Also this is the recommended way to make the vue store available to Cypress during test. When in production, the if condition prevents cypress from being injected.

File: src/main.js (hunk 15 above)
Comment (vickywane) [reply src/main.js:30]: Yes, my branch is fully synced, 

Previously the vue window was being injected into the app in this conditional statement and that was what i removed. 
E.g `window.vue = VueApp`. 

The current Vue instance is still being used (see [comment](https://github.com/ambianic/ambianic-ui/pull/581#issuecomment-886987283) ). It was there before this PR ( see [main.js](https://github.com/ambianic/ambianic-ui/blob/master/src/main.js) file in master branch )

File: src/main.js (hunk 13 above)
Comment (vickywane) [reply src/main.js:30]: I synced it and the  changes are correct. See subsequent comments.

**Comment by vickywane** [comment]
@ivelin 

If you prefer the Cypress instance removed, i might have to remove the test case dependent on it.
